### PR TITLE
Selfless Entity

### PR DESCRIPTION
**Pull request by m6jones**
Used when there is a sub-entity with no self attribute. Allows the sub-entities of these have a parent that will clean them up.

### CODE REVIEW

**Comment by ctwomey1** [comment]
Embedded siren representations are suppose to have a self link

> Root entities and sub-entities that are embedded representations SHOULD contain a links collection with at least one item contain a rel value of self and an href attribute with a value of the entity's URI.
> 
> Sub-entities that are embedded links MUST contain an href attribute with a value of its URI.
from: https://github.com/kevinswiber/siren#entities

I'm not sure we should actually do this

**Comment by m6jones** [comment]
This doesn't mean they all do though :( 

image

I rather have proper clean up on these entities.

**Comment by m6jones** [comment]
We will see about adding self links first.

File: src/activities/ActionCollectionEntity.js [hunk 1]
```diff
@@ -1,25 +1,73 @@
...
 
-import { Entity } from '../es6/Entity.js';
 import { Rels } from '../hypermedia-constants';
 import { ActivityUsageEntity } from './ActivityUsageEntity.js';
 import { performSirenAction } from '../es6/SirenAction';
 
+/**
+ * Abstract Entity class to help create entity classes.
+ */
+class SelflessEntity {
```
Comment (ctwomey1) [inline src/activities/ActionCollectionEntity.js]: seems to be a dupe of src/es6/selflessentity

File: src/es6/SelflessEntity.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
+'use strict';
+
+/**
+ * Abstract Entity class to help create entity classes.
+ */
+export class SelflessEntity {
+	/**
+	 * Primes the object used by the entityFactory. Should never be called outside.
```
Comment (ctwomey1) [inline src/es6/SelflessEntity.js]: I thought the intent was that it wasn't used by the entityfactory?

File: src/es6/SelflessEntity.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	 * @param {*} className Class name you want to see if the entity has attached to it.
+	 */
+	hasClass(className) {
+		return this._entity && this._entity.hasClass(className);
+	}
+	/**
+	 * Get the action with an Action name that is attached to the entity
+	 * @param {*} actionName Action name to get the action attached to the entity.
+	 */
+	getActionByName(actionName) {
```
Comment (ctwomey1) [inline src/es6/SelflessEntity.js]: https://github.com/BrightspaceHypermediaComponents/siren-sdk/blob/master/src/es6/Entity.js#L58-L63

The code duplication introduced here feels a bit off.  We should probably consolidate things into something logical?  Could be a simple utility? Worth thinking about

File: src/es6/SelflessEntity.js (hunk 2 above)
Comment (m6jones) [reply src/es6/SelflessEntity.js]: Oops! I updated the params but not the initial comment. :(

**Comment by ryantmer** [comment]
Yeah, lots don't - the issue is that a lot aren't addressable, and while having a self link doesn't necessarily mean they are addressable, we don't have an existing pattern to differentiate them. (For example, the due date entity on an activity usage can't be fetched directly.)

**Comment by m6jones** [comment]
Added the new pattern that is cleaned up.
Tasks:
- [ ] Add tests. Note that a test to check if the entity does not go into the store is important

**Comment by jlai-d2l** [comment]
@ctwomey1: Per our conversation, I've updated the documentation to reflect the transient nature of these entities.

**Pull request merged**